### ISSUE TITLE
Invalid client_id ("authservice-oidc") after install kubeflow on microk8s cluster-on-premise

### ISSUE THREAD

**Issue by ubntelton** [canonical/bundle-kubeflow#388]
login http://10.64.140.43.nip.io/ has the Invalid client_id ("authservice-oidc") , after install kubeflow on microk8s cluster[on-premise]
Any idea for this kind of auth issue?

![dex_bad_request](https://user-images.githubusercontent.com/61405324/129019480-7ef91fec-a22a-4128-a8a7-ed2a358d3316.JPG)

```
ai4:~$ kubectl logs pod/dex-auth-6dd645d444-pgxv6 -n kubeflow
time="2021-08-11T10:58:54Z" level=info msg="config using log level: debug"
time="2021-08-11T10:58:54Z" level=info msg="config issuer: http://10.64.140.43.nip.io/dex"
time="2021-08-11T10:58:54Z" level=info msg="kubernetes client apiVersion = dex.coreos.com/v1"
time="2021-08-11T10:58:54Z" level=info msg="creating custom Kubernetes resources"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource authcodes.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource authcodes.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource authrequests.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource authrequests.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource oauth2clients.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource oauth2clients.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource signingkeies.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource signingkeies.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource refreshtokens.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource refreshtokens.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource passwords.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource passwords.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource offlinesessionses.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource offlinesessionses.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource connectors.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource connectors.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource devicerequests.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource devicerequests.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="checking if custom resource devicetokens.dex.coreos.com has been created already..."
time="2021-08-11T10:58:54Z" level=info msg="The custom resource devicetokens.dex.coreos.com already available, skipping create"
time="2021-08-11T10:58:54Z" level=info msg="config storage: kubernetes"
time="2021-08-11T10:58:54Z" level=info msg="config connector: local passwords enabled"
time="2021-08-11T10:58:54Z" level=info msg="config skipping approval screen"
time="2021-08-11T10:58:54Z" level=info msg="listening (http) on 0.0.0.0:5556"
time="2021-08-11T11:06:39Z" level=error msg="Failed to parse authorization request: Invalid client_id (\"authservice-oidc\")."
time="2021-08-11T11:06:49Z" level=error msg="Failed to parse authorization request: **Invalid client_id (\"authservice-oidc\")."**
```

**Comment by Pedrohgv** [comment]
Having the same issue. Funny it worked on the first try, and after restart stoped working. Tried reinstalling both kubeflow deployment and microk8s but the error persisted.

**Comment by edhenry** [comment]
Was able to resolve this, myself, by deleting the `oidc-gatekeeper` pod and letting the respective operator re-deploy it. 

```bash
microk8s.kubectl delete pod -n kubeflow `<oidc-gatekeeper-pod-id>`
```

As mentioned above the pod should be redeployed by the `oidc-gatekeeper-operator` pod that is deployed.

Hopefully this will be helpful for others hitting this issue!

**Comment by DnPlas** [comment]
Hi @ubntelton thanks for reporting this. After trying a couple of times, I was not able to reproduce the issue. I could see a similar behaviour under very specific conditions, but I had to manually make it fail.

Just for reference, you'll get `Invalid client_id ("authservice-oidc")` when there is a mismatch between `staticClients` and `clientID` (CLIENT_ID) in `dex-auth`'s and `oidc-gatekeeper`'s configMaps respectively. These values are setup by the charm code pretty much on every event handler, which should ensure correctness, and explains why after deleting the `oidc-gatekeeper-*` the problem is solved: it triggers some hooks and re-runs the piece of code that updates those configMaps.

If you are still running into the issue, could you please share the output of `microk8s.kubectl get configmap/dex-auth-config-config -n kubeflow -o yaml` and `juju status --relations`. Also, how are you deploying kubeflow? @banderlog do you mind also sharing this?

#### My setup
microk8s v1.21.5
Ubuntu 20.04.3 LTS

#### Steps
Basically  [this](https://www.kubeflow.org/docs/distributions/microk8s/kubeflow-on-microk8s/)

Kubeflow:
$ microk8s status --wait-ready
$ microk8s enable kubeflow
Wait...
[I was able to login with credentials from the output, login -> 10.64.140.43.nip.io/]

Kubeflow lite:
$ microk8s status --wait-ready
$ microk8s enable kubeflow --bundle=cs:kubeflow-lite
Wait...
[I was able to login with credentials from the output, login -> 10.64.140.43.nip.io/]

**Comment by banderlog** [comment]
Well, I just repeated all steps that I did before (using [official documentation](https://www.kubeflow.org/docs/distributions/microk8s/kubeflow-on-microk8s/)):

```
sudo snap install microk8s --classic --channel=1.21/stable
microk8s status --wait-ready
microk8s enable kubeflow
# wait 35min
```

Really, nothing changed in my setup or installation steps, but this time I was able to login with credentials from the output.

**Setup:**
microk8s v1.21.5
Ubuntu 20.04.3 LTS

**Comment by yingding** [comment]
Thanks @edhenry 

> Was able to resolve this, myself, by deleting the oidc-gatekeeper pod and letting the respective operator re-deploy it.
> ```
> microk8s.kubectl delete pod -n kubeflow <oidc-gatekeeper-pod-id>
> ```

On my Ubuntu 20.04 with MicroK8s 1.21.6, i just delete the pod
```
microk8s.kubectl delete pod -n kubeflow oidc-gatekeeper-xxxxxxx-xvm6g
```
and the pod is redeployed automatically. I can then login fine.

**Comment by grobbie** [comment]
> microk8s.kubectl delete pod -n kubeflow oidc-gatekeeper-xxxxxxx-xvm6g

This trick worked for me as well, although I had to wait for a while for juju to bring everything back into order:
`watch -c juju --color`

**Comment by DomFleischmann** [comment]
This should be fixed with the latest release of Charmed Kubeflow. Feel free to reopen if it still happens.
